From 326bf28498cc37c2b6d7ec48df6f46546fa37877 Mon Sep 17 00:00:00 2001
From: "Nguyen Van Tu [FGA.AIS]" <TuNV16@fsoft.com.vn>
Date: Fri, 19 Oct 2018 14:49:57 +0700
Subject: [PATCH 200/642] ARM: shmobile: Initial r8a77470 SoC support

Add initial support for the r8a77470 SoC including:
 - 2 Cortex-A7 CPU Core
 - GIC

Signed-off-by: thongsyho <thong.ho.px@rvc.renesas.com>
Signed-off-by: Nguyen Van Linh [FGA.BU13] <LinhNV25@fsoft.com.vn>
Signed-off-by: Viet Nguyen <vietn@fsoft.com.vn>
Signed-off-by: vietn <vietn@fsoft.com.vn>
Signed-off-by: thongsyho <thong.ho.px@rvc.renesas.com>
Signed-off-by: Nguyen Van Tu [FGA.AIS] <TuNV16@fsoft.com.vn>
---
 arch/arm/boot/dts/r8a77470.dtsi          | 55 ++++++++++++++++++++++++++++++++
 arch/arm/mach-shmobile/Kconfig           |  5 +++
 arch/arm/mach-shmobile/Makefile          |  2 ++
 arch/arm/mach-shmobile/pm-rcar-gen2.c    |  2 ++
 arch/arm/mach-shmobile/r8a77470.h        |  6 ++++
 arch/arm/mach-shmobile/setup-r8a77470.c  | 38 ++++++++++++++++++++++
 arch/arm/mach-shmobile/setup-rcar-gen2.c |  9 ++++--
 arch/arm/mach-shmobile/smp-r8a77470.c    | 53 ++++++++++++++++++++++++++++++
 8 files changed, 168 insertions(+), 2 deletions(-)
 create mode 100644 arch/arm/boot/dts/r8a77470.dtsi
 create mode 100644 arch/arm/mach-shmobile/r8a77470.h
 create mode 100644 arch/arm/mach-shmobile/setup-r8a77470.c
 create mode 100644 arch/arm/mach-shmobile/smp-r8a77470.c

diff --git a/arch/arm/boot/dts/r8a77470.dtsi b/arch/arm/boot/dts/r8a77470.dtsi
new file mode 100644
index 0000000..83866b2
--- /dev/null
+++ b/arch/arm/boot/dts/r8a77470.dtsi
@@ -0,0 +1,55 @@
+/*
+ * Device Tree Source for the r8a77470 SoC
+ *
+ * Copyright (C) 2013-2015 Renesas Electronics Corporation
+ *
+ * This file is licensed under the terms of the GNU General Public License
+ * version 2.  This program is licensed "as is" without any warranty of any
+ * kind, whether express or implied.
+ */
+
+#include <dt-bindings/interrupt-controller/arm-gic.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+
+/ {
+	compatible = "renesas,r8a77470";
+	interrupt-parent = <&gic>;
+	#address-cells = <2>;
+	#size-cells = <2>;
+
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu0: cpu@0 {
+			device_type = "cpu";
+			compatible = "arm,cortex-a7";
+			reg = <0>;
+			clock-frequency = <1000000000>;
+			voltage-tolerance = <1>; /* 1% */
+			clock-latency = <300000>; /* 300 us */
+
+			/* kHz - uV - OPPs unknown yet */
+			operating-points = <1000000 1000000>;
+		};
+
+		cpu1: cpu@1 {
+			device_type = "cpu";
+			compatible = "arm,cortex-a7";
+			reg = <1>;
+			clock-frequency = <1000000000>;
+		};
+	};
+
+	gic: interrupt-controller@f1001000 {
+		compatible = "arm,cortex-a7-gic";
+		#interrupt-cells = <3>;
+		#address-cells = <0>;
+		interrupt-controller;
+		reg = <0 0xf1001000 0 0x1000>,
+			<0 0xf1002000 0 0x1000>,
+			<0 0xf1004000 0 0x2000>,
+			<0 0xf1006000 0 0x2000>;
+		interrupts = <1 9 (GIC_CPU_MASK_SIMPLE(4) | IRQ_TYPE_LEVEL_HIGH)>;
+	};
+};
\ No newline at end of file
diff --git a/arch/arm/mach-shmobile/Kconfig b/arch/arm/mach-shmobile/Kconfig
index 107a6b4..bec7ad1 100644
--- a/arch/arm/mach-shmobile/Kconfig
+++ b/arch/arm/mach-shmobile/Kconfig
@@ -75,6 +75,11 @@ config ARCH_R8A7745
 	bool "RZ/G1E (R8A77450)"
 	select ARCH_RCAR_GEN2
 
+config ARCH_R8A77470
+	bool "RZ/G1C (R8A77470)"
+	select ARCH_RCAR_GEN2
+	select I2C
+
 config ARCH_R8A7778
 	bool "R-Car M1A (R8A77781)"
 	select ARCH_RCAR_GEN1
diff --git a/arch/arm/mach-shmobile/Makefile b/arch/arm/mach-shmobile/Makefile
index 02824b7..c8ed175 100644
--- a/arch/arm/mach-shmobile/Makefile
+++ b/arch/arm/mach-shmobile/Makefile
@@ -14,6 +14,7 @@ obj-$(CONFIG_ARCH_R8A7779)	+= setup-r8a7779.o pm-r8a7779.o
 obj-$(CONFIG_ARCH_R8A7790)	+= setup-r8a7790.o
 obj-$(CONFIG_ARCH_R8A7791)	+= setup-r8a7791.o
 obj-$(CONFIG_ARCH_R8A7745)      += setup-r8a7745.o sgx-r8a7745-quirk.o
+obj-$(CONFIG_ARCH_R8A77470)     += setup-r8a77470.o
 obj-$(CONFIG_ARCH_EMEV2)	+= setup-emev2.o
 obj-$(CONFIG_ARCH_R7S72100)	+= setup-r7s72100.o
 
@@ -34,6 +35,7 @@ smp-$(CONFIG_ARCH_SH73A0)	+= smp-sh73a0.o headsmp-scu.o platsmp-scu.o
 smp-$(CONFIG_ARCH_R8A7779)	+= smp-r8a7779.o headsmp-scu.o platsmp-scu.o
 smp-$(CONFIG_ARCH_R8A7790)	+= smp-r8a7790.o
 smp-$(CONFIG_ARCH_R8A7791)	+= smp-r8a7791.o
+smp-$(CONFIG_ARCH_R8A77470)     += smp-r8a77470.o
 smp-$(CONFIG_ARCH_EMEV2)	+= smp-emev2.o headsmp-scu.o platsmp-scu.o
 
 # PM objects
diff --git a/arch/arm/mach-shmobile/pm-rcar-gen2.c b/arch/arm/mach-shmobile/pm-rcar-gen2.c
index bead25e..0b9bb93 100644
--- a/arch/arm/mach-shmobile/pm-rcar-gen2.c
+++ b/arch/arm/mach-shmobile/pm-rcar-gen2.c
@@ -96,6 +96,8 @@ void __init rcar_gen2_pm_init(void)
 		syscier = 0x00111003;
 	else if (of_machine_is_compatible("renesas,r8a7743"))
 		syscier = 0x00101003;
+	else if (of_machine_is_compatible("renesas,r8a77470"))
+		syscier = 0x0131000e;
 	else if (of_machine_is_compatible("renesas,r8a7745"))
 		syscier = 0x00300060;
 
diff --git a/arch/arm/mach-shmobile/r8a77470.h b/arch/arm/mach-shmobile/r8a77470.h
new file mode 100644
index 0000000..74b9fa3
--- /dev/null
+++ b/arch/arm/mach-shmobile/r8a77470.h
@@ -0,0 +1,6 @@
+#ifndef __ASM_R8A77470_H__
+#define __ASM_R8A77470_H__
+
+extern struct smp_operations r8a77470_smp_ops;
+
+#endif /* __ASM_R8A77470_H__ */
\ No newline at end of file
diff --git a/arch/arm/mach-shmobile/setup-r8a77470.c b/arch/arm/mach-shmobile/setup-r8a77470.c
new file mode 100644
index 0000000..a70e5f8
--- /dev/null
+++ b/arch/arm/mach-shmobile/setup-r8a77470.c
@@ -0,0 +1,38 @@
+/*
+ * r8a77470 processor support
+ *
+ * Copyright (C) 2013  Renesas Electronics Corporation
+ * Copyright (C) 2013  Renesas Solutions Corp.
+ * Copyright (C) 2013  Magnus Damm
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include <linux/init.h>
+
+#include <asm/mach/arch.h>
+
+#include "common.h"
+#include "r8a77470.h"
+#include "rcar-gen2.h"
+
+static const char *const r8a77470_boards_compat_dt[] __initconst = {
+	"renesas,r8a77470",
+	NULL,
+};
+
+DT_MACHINE_START(R8A77470_DT, "Generic R8A77470 (Flattened Device Tree)")
+	.smp		= smp_ops(r8a77470_smp_ops),
+	.init_early	= shmobile_init_delay,
+	.init_time	= rcar_gen2_timer_init,
+	.init_late	= shmobile_init_late,
+	.reserve	= rcar_gen2_reserve,
+	.dt_compat	= r8a77470_boards_compat_dt,
+MACHINE_END
\ No newline at end of file
diff --git a/arch/arm/mach-shmobile/setup-rcar-gen2.c b/arch/arm/mach-shmobile/setup-rcar-gen2.c
index 209b8fa..31f3519 100644
--- a/arch/arm/mach-shmobile/setup-rcar-gen2.c
+++ b/arch/arm/mach-shmobile/setup-rcar-gen2.c
@@ -61,7 +61,8 @@ void __init rcar_gen2_timer_init(void)
 	shmobile_init_cntvoff();
 
 	if (of_machine_is_compatible("renesas,r8a7745") ||
-	    of_machine_is_compatible("renesas,r8a7794")) {
+	    of_machine_is_compatible("renesas,r8a7794") ||
+	    of_machine_is_compatible("renesas,r8a77470")) {
 		freq = 260000000 / 8;	/* ZS / 8 */
 	} else {
 		/* At Linux boot time the r8a7790 arch timer comes up
@@ -177,7 +178,11 @@ void __init rcar_gen2_reserve(void)
 
 	/* reserve 256 MiB at the top of the physical legacy 32-bit space */
 	memset(&mrc, 0, sizeof(mrc));
-	mrc.reserved = SZ_256M;
+
+	if (of_machine_is_compatible("renesas,r8a77470"))
+		mrc.reserved = SZ_128M;
+	else
+		mrc.reserved = SZ_256M;
 
 	of_scan_flat_dt(rcar_gen2_scan_mem, &mrc);
 #ifdef CONFIG_DMA_CMA
diff --git a/arch/arm/mach-shmobile/smp-r8a77470.c b/arch/arm/mach-shmobile/smp-r8a77470.c
new file mode 100644
index 0000000..9d870da
--- /dev/null
+++ b/arch/arm/mach-shmobile/smp-r8a77470.c
@@ -0,0 +1,53 @@
+/*
+ * SMP support for r8a77470
+ *
+ * Copyright (C) 2013 Renesas Solutions Corp.
+ * Copyright (C) 2013 Magnus Damm
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+#include <linux/kernel.h>
+#include <linux/init.h>
+#include <linux/smp.h>
+#include <linux/io.h>
+
+#include <asm/smp_plat.h>
+
+#include "common.h"
+#include "platsmp-apmu.h"
+#include "r8a77470.h"
+#include "rcar-gen2.h"
+
+static struct rcar_apmu_config r8a77470_apmu_config[] = {
+	{
+		.iomem = DEFINE_RES_MEM(0xe6152000, 0x188),
+		.cpus = { 0, 1 },
+	}
+};
+
+static void __init r8a77470_smp_prepare_cpus(unsigned int max_cpus)
+{
+	/* let APMU code install data related to shmobile_boot_vector */
+	shmobile_smp_apmu_prepare_cpus(max_cpus,
+				       r8a77470_apmu_config,
+				       ARRAY_SIZE(r8a77470_apmu_config));
+
+	rcar_gen2_pm_init();
+}
+
+struct smp_operations r8a77470_smp_ops __initdata = {
+	.smp_prepare_cpus	= r8a77470_smp_prepare_cpus,
+	.smp_boot_secondary	= shmobile_smp_apmu_boot_secondary,
+#ifdef CONFIG_HOTPLUG_CPU
+	.cpu_can_disable	= shmobile_smp_cpu_can_disable,
+	.cpu_die		= shmobile_smp_apmu_cpu_die,
+	.cpu_kill		= shmobile_smp_apmu_cpu_kill,
+#endif
+};
\ No newline at end of file
-- 
2.7.4

